File: ios/MrzKey.swift
import Foundation
import CryptoKit

enum MrzKey {

    private static let weights = [7, 3, 1]

    static func checkDigit(_ input: String) -> Character {
        var sum = 0
        for (i, c) in input.enumerated() {
            let v: Int
            if c.isNumber { v = c.wholeNumberValue ?? 0 }
            else if c == "<" { v = 0 }
            else if let a = c.asciiValue, c.isLetter { v = Int(a) - 65 + 10 }
            else { v = 0 }
            sum += v * weights[i % 3]
        }
        return Character(String(sum % 10))
    }

    static func mrzInformation(doc: String, dob: String, expiry: String) -> String {
        var d = doc.uppercased()
        while d.count < 9 { d += "<" }
        return d + String(checkDigit(d)) + dob + String(checkDigit(dob))
             + expiry + String(checkDigit(expiry))
    }

    /// يرجع (Kenc, Kmac)
    static func deriveBacKeys(doc: String, dob: String, expiry: String) -> (Data, Data) {
        let info = mrzInformation(doc: doc, dob: dob, expiry: expiry)
        let seed = Data(Insecure.SHA1.hash(data: Data(info.utf8)).prefix(16))
        return (kdf(seed: seed, counter: 1), kdf(seed: seed, counter: 2))
    }

    static func kdf(seed: Data, counter: UInt8) -> Data {
        let d = seed + Data([0, 0, 0, counter])
        var key = Array(Insecure.SHA1.hash(data: d).prefix(16))
        for i in key.indices {
            let ones = (key[i] & 0xFE).nonzeroBitCount
            key[i] = (ones % 2 == 0) ? (key[i] | 0x01) : (key[i] & 0xFE)
        }
        return Data(key)
    }
}

File: android/src/main/java/expo/modules/nfciraqireader/nfc/MrzKey.kt
package expo.modules.nfciraqireader.nfc

import java.security.MessageDigest

/**
 * اشتقاق مفاتيح BAC من بيانات الـ MRZ المطبوعة على وجه/ظهر البطاقة.
 * هاي نقطة مهمة: بدون البيانات المطبوعة ما تكدر تفتح الشريحة — هذا مقصود بالتصميم.
 */
object MrzKey {

    private val WEIGHTS = intArrayOf(7, 3, 1)

    fun checkDigit(input: String): Char {
        var sum = 0
        input.forEachIndexed { i, c ->
            val v = when {
                c in '0'..'9' -> c - '0'
                c in 'A'..'Z' -> c - 'A' + 10
                c == '<' -> 0
                else -> throw IllegalArgumentException("حرف غير مسموح: $c")
            }
            sum += v * WEIGHTS[i % 3]
        }
        return '0' + (sum % 10)
    }

    /** docNumber = رقم الوثيقة، dob و expiry بصيغة YYMMDD */
    fun mrzInformation(docNumber: String, dob: String, expiry: String): String {
        val doc = docNumber.uppercase().padEnd(9, '<')
        require(doc.length == 9) { "رقم الوثيقة أطول من 9 خانات — راجع ملاحظات README" }
        require(dob.length == 6 && expiry.length == 6) { "التاريخ لازم يكون YYMMDD" }
        return doc + checkDigit(doc) + dob + checkDigit(dob) + expiry + checkDigit(expiry)
    }

    /** يرجع (Kenc, Kmac) */
    fun deriveBacKeys(docNumber: String, dob: String, expiry: String): Pair<ByteArray, ByteArray> {
        val info = mrzInformation(docNumber, dob, expiry)
        val seed = MessageDigest.getInstance("SHA-1")
            .digest(info.toByteArray(Charsets.US_ASCII))
            .copyOf(16)
        return Pair(kdf(seed, 1), kdf(seed, 2))
    }

    /** counter = 1 للتشفير، 2 للـ MAC */
    fun kdf(seed: ByteArray, counter: Int): ByteArray {
        val d = seed + byteArrayOf(0, 0, 0, counter.toByte())
        val key = MessageDigest.getInstance("SHA-1").digest(d).copyOf(16)
        adjustParity(key)
        return key
    }

    private fun adjustParity(key: ByteArray) {
        for (i in key.indices) {
            val b = key[i].toInt() and 0xFF
            key[i] = if (Integer.bitCount(b and 0xFE) % 2 == 0) (b or 0x01).toByte()
                     else (b and 0xFE).toByte()
        }
    }
}
